### ISSUE TITLE
Document usage of Pytorch half() when loading model

### ISSUE THREAD

**Issue by kingpalethe** [guillaume-be/rust-bert#183]
Thanks for this great package.

For some of the bigger models used by this library, like GPT-NEO-2.7, the GPU memory requirements are quite large.

I would expect therefore that many users of this library might want to use convert these models  to `fp16` before loading onto a GPU, using `.half()`.

For example, in Python transformers, I can load GPT-NEO-2.7 for inference onto an 8GB rtx 2080 without problems, as long as I use `half()`, like this:

```
model = GPTNeoForCausalLM.from_pretrained(
            model_name).half().to(torch_device)
```

My guess is that this is already possible with `tch-rs`, but I am struggling to find the right syntax to do this. I actually see that @guillaume-be added similar functionality to `tch-rs` in a June 2020 pull request here: https://github.com/LaurentMazare/tch-rs/pull/221/commits/0c50286189a7fb707f1e1822dc06e707a15a94b8 , and I see the `Kind` documentation here: https://docs.rs/tch/0.5.0/tch/kind/enum.Kind.html .

My guess is some kind call to `Kind::Half` would need to be applied to this line: https://github.com/guillaume-be/rust-bert/blob/13995b35415277ab5ae80a6adeeef99b5f7fdd8d/examples/generation_gpt_neo.rs#L36  -- but I'm at a loss on how to do it.

Thanks.

**Comment by kingpalethe** [comment]
Thanks for this update.

**Comment by guillaume-be** [comment]
Hello @kingpalethe ,

I wanted to share an update on the status of this issue. I have now a working branch with an implementation of half-precision conversion for all models implemented in this library. This relies on a new mechanism introduced in the [tch](https://github.com/LaurentMazare/tch-rs) crate allowing to move a variable store between devices and change their type. The operations you would perform on the `nn.Module` in Pytorch are now performed on the `VarStore` containing the model parameters. I have exposed a similar API for the pipelines - see for example for GPT-Neo:
```rust
fn main() -> anyhow::Result<()> {
    //    Set-up model resources
    let config_resource = Resource::Remote(RemoteResource::from_pretrained(
        GptNeoConfigResources::GPT_NEO_125M,
    ));
    let vocab_resource = Resource::Remote(RemoteResource::from_pretrained(
        GptNeoVocabResources::GPT_NEO_125M,
    ));
    let merges_resource = Resource::Remote(RemoteResource::from_pretrained(
        GptNeoMergesResources::GPT_NEO_125M,
    ));
    let model_resource = Resource::Remote(RemoteResource::from_pretrained(
        GptNeoModelResources::GPT_NEO_125M,
    ));
    let generate_config = TextGenerationConfig {
        model_type: ModelType::GPTNeo,
        model_resource,
        config_resource,
        vocab_resource,
        merges_resource,
        min_length: 10,
        max_length: 32,
        do_sample: false,
        early_stopping: true,
        num_beams: 4,
        num_return_sequences: 1,
        device: Device::Cpu,
        ..Default::default()
    };

    let mut model = TextGenerationModel::new(generate_config)?;
    model.half();
    model.set_device(Device::cuda_if_available());

    let input_context_1 = "It was a very nice and sunny";
    let input_context_2 = "It was a gloom winter night, and";
    let output = model.generate(&[input_context_1, input_context_2], None);

    for sentence in output {
        println!("{}", sentence);
    }
    Ok(())
}
```

This will load the model to CPU, convert it to `fp16` precision and move it to the GPU. 
We need to wait for the changes introduced in [tch](https://github.com/LaurentMazare/tch-rs) in https://github.com/LaurentMazare/tch-rs/pull/416 and https://github.com/LaurentMazare/tch-rs/pull/420 to be released before merging these changes, I will let you know as soon as this is ready for use.

**Comment by kingpalethe** [comment]
Great, looking forward to trying this when it's ready. Thanks.

**Comment by guillaume-be** [comment]
Hello @kingpalethe ,

I just opened a PR that would implement the featured discussed now that a new version of `tch` has been released. This allows converting the models either directly on the pipeline, or using the `VarStore` `half()` method. These changes will be merged into master today, and should be released by the end of this week.
